### DISCUSSION TITLE
Exclusion of values when calculating coefficients

### DISCUSSION THREAD

**Discussion by Aurilino** [casact/chainladder-python#383] [Q&A]
Hello!

It can help, there are a lot of zeros in my triangle and when calculating the development coefficients of the value, if the value in the period comes after the zero value in the previous period, then it is excluded from the sum for the period. I need the sum of the entire column without exceptions to be taken to calculate the coefficient.
![image](https://user-images.githubusercontent.com/47794083/203843447-f421e17e-0db7-46f1-b701-28a2717d913d.png)

**Comment by kennethshsu** [comment]
Hi @Aurilino, welcome!

I don't think it makes sense to include $13,255 for the calculation of the coefficient in your example, the same reasoning can be applied when excluding $18,632 on the first column - because these aren't really loss developments.

Now, do you have a $0 for the cell to the left of $13,255? I think $13,255 should only be included if it developed from $0, and not null, from missing data. Do you agree?

**Comment by jbogaardt** [comment]
For a volume-weighted deterministic approach, I can see why this might be useful. Some lines with late reporting, like construction defect might have zeros for earlier entries and this is useful information for estimating patterns.  It wouldn't make sense for simple average link ratios because of division by zero issues.  

The `Development` estimator as written does not support this. It is based on the regression framework proposed by Mack which (even for volume-weighted) also runs into division by zero issues here.  This is an interesting case where the desired outcome defies the framework. While it is a trivial calculation, I struggle with incorporating it with any elegance.  A non-elegant but practical approach for a user could employ would be to add an infinitesimal amount to the triangle.
```python
triangle + 1e-16
```
This would get your entry to show up in the weighted average but be an insignificant bias in the estimation process.  That said, I wouldn't rely on the stochastic properties (`std_err_`, `sigma_`) of the estimator if doing this. It seems fit only fore purely determinisitic reserving.

**Reply by Aurilino** [reply]
Thanks
I thought about that too, but was hoping there was a built-in solution.
When reporting, my regulator requires that we take the sum of the column and divide it by the sum of the previous column minus the last period.
For myself, I came up with the following solution for calculating the coefficients:
`diagonal = np.diag(triangle.loc[:,::-1])`
`ldf = triangle.sum().shift(-1) / (triangle * abs(np.eye(len(triangle))-1)`
`cdf = [pd.Series[j for j in ldf[i,::]).prod(min_count=1) for i in range(len(ldf)]`
